### PR TITLE
 convert latency/latencies in metrics name to duration

### PR DESCRIPTION
**Pull request by danielqsj**
**What type of PR is this?**

/kind feature
/sig instrumentation

**What this PR does / why we need it**:

As discussion [here](https://github.com/kubernetes/kubernetes/pull/72470#discussion_r257612378) , we need to convert `latency`/`latencies` in metrics name to `duration` to follow the best practice for Prometheus.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Convert `latency`/`latencies` in metrics name to `duration`.
The following metrics are changed and mark previous metrics as deprecated:
* `rest_client_request_latency_seconds` -> `rest_client_request_duration_seconds`
* `apiserver_proxy_tunnel_sync_latency_secs` -> `apiserver_proxy_tunnel_sync_duration_seconds`
* `scheduler_scheduling_latency_seconds` -> `scheduler_scheduling_duration_seconds `
```

### CODE REVIEW

File: pkg/util/workqueue/prometheus/prometheus.go [hunk 1]
```diff
@@ -31,7 +31,7 @@ const (
 	WorkQueueSubsystem         = "workqueue"
 	DepthKey                   = "depth"
 	AddsKey                    = "adds_total"
-	QueueLatencyKey            = "queue_latency_seconds"
+	QueueLatencyKey            = "queue_duration_seconds"
```
Comment (brancz) [inline pkg/util/workqueue/prometheus/prometheus.go]: I don't see deprecation for this?

**Review by brancz — COMMENTED** [review]
One question, otherwise I double checked that we have already deprecated the old metrics for those changed here, so we're good to go once we can confirm this one other one.

File: pkg/util/workqueue/prometheus/prometheus.go (hunk 1 above)
Comment (danielqsj) [reply pkg/util/workqueue/prometheus/prometheus.go]: This metrics https://github.com/kubernetes/kubernetes/blob/c525d329effc6c6460cda947d1bf8092a927c2d3/pkg/util/workqueue/prometheus/prometheus.go#L69 is new introduced in 1.14 by https://github.com/kubernetes/kubernetes/pull/71300, to deprecate this metrics`queue_latency` https://github.com/kubernetes/kubernetes/blob/c525d329effc6c6460cda947d1bf8092a927c2d3/pkg/util/workqueue/prometheus/prometheus.go#L155

Comment (danielqsj) [reply pkg/util/workqueue/prometheus/prometheus.go]: What I think is the newly introduced metrics (via KEP, but name contains `latency/latencies`)  will not need to be replaced by another newer metrics (replace the name with `duration`), by this change, right ?

Comment (brancz) [reply pkg/util/workqueue/prometheus/prometheus.go]: Yes you’re totally right I just didn’t see it :)

**Review by liggitt — COMMENTED** [review]
I'll defer to @kubernetes/sig-instrumentation-api-reviews on the cost of doubling these up during the deprecation period

File: pkg/master/master.go [hunk 2]
```diff
@@ -387,9 +387,13 @@ func (m *Master) installTunneler(nodeTunneler tunneler.Tunneler, nodeClient core
 	nodeTunneler.Run(nodeAddressProvider{nodeClient}.externalAddresses)
 	m.GenericAPIServer.AddHealthzChecks(healthz.NamedCheck("SSH Tunnel Check", tunneler.TunnelSyncHealthChecker(nodeTunneler)))
 	prometheus.NewGaugeFunc(prometheus.GaugeOpts{
-		Name: "apiserver_proxy_tunnel_sync_latency_secs",
+		Name: "apiserver_proxy_tunnel_sync_duration_seconds",
 		Help: "The time since the last successful synchronization of the SSH tunnels for proxy requests.",
 	}, func() float64 { return float64(nodeTunneler.SecondsSinceSync()) })
+	prometheus.NewGaugeFunc(prometheus.GaugeOpts{
+		Name: "apiserver_proxy_tunnel_sync_latency_secs",
+		Help: "(Deprecated) The time since the last successful synchronization of the SSH tunnels for proxy requests.",
```
[the remark this answers is not in the document]
Comment (danielqsj) [reply pkg/master/master.go]: This change is part of the KEP [Kubernetes Metrics Overhaul](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/0031-kubernetes-metrics-overhaul.md#deprecation-plan) . We plan to mark all deprecated metrics in 1.14, and remove them in 1.15.
The pattern "(Deprecated)" in front of the metric help text is consistent with other changes.
Does it sound good ?
/cc @brancz

[the remark this answers is not in the document]
Comment (brancz) [reply pkg/master/master.go]: There are no guarantees around metrics at all right now, so technically we can break as much as we want, but we decided we want at least 1 release as transition period.

**Comment by brancz** [comment]
I think for high cardinality metrics we can decide on a case by case basis which metrics we may not be ok with duplicating. For what it's worth, Prometheus has a built-in mechanism to drop metrics at ingestion, so apart from network bytes transferred, there is no efficiency loss, as dropping metrics at ingestion is a very cheap operation. As described in the KEP, we are providing transition mechanisms and lists of metrics before/after so that people can convert their setups in the 1.14 timeframe while maintaining efficiency of their Prometheus setups.

**Comment by roycaihw** [comment]
cc @logicalhan

**Comment by danielqsj** [comment]
ping @liggitt

**Comment by brancz** [comment]
cc @smarterclayton

**Comment by liggitt** [comment]
> As described in the KEP, we are providing transition mechanisms and lists of metrics before/after so that people can convert their setups in the 1.14 timeframe while maintaining efficiency of their Prometheus setups.

Should the changed names be described in detail in the PR release note?

File: pkg/master/master.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (liggitt) [reply pkg/master/master.go]: for visibility, asked for an ack on that position [from sig-instrumentation and API reviewers](https://groups.google.com//forum/#!topic/kubernetes-api-reviewers/kdxON6-TtNE). can proceed once there's agreement there

**Comment by danielqsj** [comment]
@liggitt Updated release note with all metrics changed here.

**Comment by brancz** [comment]
Absolutely, thanks for the catch @liggitt!

**Comment by liggitt** [comment]
> The following metrics are new introduced in 1.14, so they are directly changed:
...
> * `*_admission_latencies_seconds_summary` -> `*_admission_duration_seconds_summary`

these two existed in 1.13:

https://github.com/kubernetes/kubernetes/blob/release-1.13/staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go#L160-L189

File: pkg/proxy/metrics/metrics.go [hunk 3]
```diff
@@ -56,13 +56,25 @@ var (
...
 			Buckets: prometheus.ExponentialBuckets(0.001, 2, 20),
 		},
 	)
+
+	// DeprecatedNetworkProgrammingLatency is deprecated, please use NetworkProgrammingLatency.
+	DeprecatedNetworkProgrammingLatency = prometheus.NewHistogram(
+		prometheus.HistogramOpts{
+			Subsystem: kubeProxySubsystem,
+			Name:      "network_programming_latency_seconds",
+			Help:      "(Deprecated) In Cluster Network Programming Latency in seconds",
```
Comment (liggitt) [inline pkg/proxy/metrics/metrics.go]: network_programming_latency_seconds is new in 1.14, can drop this

**Comment by liggitt** [comment]
I swept the metrics from 1.13 modified in this PR, with these results:

in 1.13:
rest_client_request_latency_seconds (deprecated)
apiserver_proxy_tunnel_sync_latency_secs (deprecated)
scheduling_latency_seconds (deprecated)
*_admission_latencies_seconds_summary (removed!)
*_admission_latencies_seconds (removed!)

new in 1.14, free to change:
docker_operations_latency_seconds
network_plugin_operations_latency_seconds
sync_proxy_rules_latency_seconds
network_programming_latency_seconds
sync_proxy_rules_latency_seconds
e2e_scheduling_latency_seconds
scheduling_algorithm_latency_seconds
binding_latency_seconds
queue_latency_seconds
apiserver_request_latency_seconds
etcd_request_cache_get_latency_seconds
etcd_request_cache_add_latency_seconds
etcd_request_latency_seconds
transformation_latencies_seconds
data_key_generation_latencies_seconds

we should restore/deprecate these two:
*_admission_latencies_seconds_summary (removed!)
*_admission_latencies_seconds (removed!)

and we can drop the new-in-1.14 metrics from the release note... the intra-release name change is just noise in the release note

**Comment by brancz** [comment]
@liggitt the admission latency metrics were intentionally removed because they had too high cardinality (not uncommon to cause 30k+ active time-series per apiserver instance).

**Comment by liggitt** [comment]
> @liggitt the admission latency metrics were intentionally removed because they had too high cardinality (not uncommon to cause 30k+ active time-series per apiserver instance).

ah, I see. can we call those out explicitly in the release note, then, along with the reason for removal?

**Comment by brancz** [comment]
Yes absolutely.

**Comment by liggitt** [comment]
also, it looks like those were renamed, not removed... are you saying the cardinality was too high to maintain the same metric under the deprecated name?

**Comment by brancz** [comment]
I recall them being reported to have high cardinality (and am seeing that in my environments currently), so my assumption here was that they were removed for that reason, but need to double check.

**Comment by danielqsj** [comment]
@liggitt  
Yes, in 1.13, these two metrics `_admission_latencies_seconds_summary` 
 and `_admission_latencies_seconds` are actually wrong in unit. And been fixed in 1.14 https://github.com/kubernetes/kubernetes/pull/72343. 
For backward compatible, We renamed these metrics which has wrong unit implements to `_admission_latencies_milliseconds_summary` and `_admission_latencies_milliseconds`.

**Comment by danielqsj** [comment]
@liggitt  @brancz 
Thanks for the finding about `network_programming_latency_seconds`.
I dropped 1.14 change in release notes.
And opened a PR to clean the deprecated metrics `network_programming_latency_seconds`. PTAL

**Comment by brancz** [comment]
Aha yes thanks for digging that up @danielqsj. For what it's worth we do drop those metrics in our setup due to being too high cardinality.

@liggitt we can do

1) leave as is as the old metric was inconsistent in its labeling (it said "seconds" but reported "milliseconds"), so this was deprecation and bugfix in one go
2) immediately remove the deprecated metric as it was both incorrectly saying what unit it is, and otherwise violated the naming scheme
3) revert the deprecated metric to the old inconsistent state, and leave deprecated and remove with other deprecated metrics
4) one of the above + look into reducing cardinality 
5) remove both metrics due to high cardinality

3 might be the least breaking thing for people making use of this metric today, 2 is what we typically do in such a case.

Looking through the most popular dashboard and alert definitions[1], none of them seem to make use of this metric and most of them even drop it. Given the nature of the metric I think it could be useful in a general sense, so I feel we should do drop the old metric entirely (2) and look into reducing cardinality with the new metric (4).

[1] https://github.com/coreos/prometheus-operator/tree/master/contrib/kube-prometheus/manifests
https://github.com/kubernetes-monitoring/kubernetes-mixin
https://github.com/grafana/jsonnet-libs/tree/master/prometheus-ksonnet/lib https://github.com/helm/charts/tree/master/stable/prometheus-operator

**Comment by brancz** [comment]
Looks like we already solved the cardinality explosion which is why we didn't remove the metric (RE https://github.com/kubernetes/kubernetes/issues/69540; I just checked it's present in 1.12 and not 1.13). In that case, we should just remove the old metric and explicitly call this out in the release notes for 1.14. Sounds good?

**Comment by danielqsj** [comment]
@brancz thanks for clarification.
I agree to directly remove the old metrics as @liggitt mentions they don't seem to help much.
I submit a PR to remove them in https://github.com/kubernetes/kubernetes/pull/75279 . PTAL

**Pull request merged**